userspace/libscap/plugin_info.h

```diff
@@ -330,297 +302,192 @@ typedef struct
+		//   array.
+		//   Each field entry is a json object with the following properties:
+		//     "name": a string with a name for the field
+		//     "type": one of "string", "uint64"
+		//     "isList: (optional) If present and set to true, notes
+		//              that the field extracts a list of values.
+		//     "argRequired": [DEPRECATED, use "arg" property instead]
+		//                   (optional) If present and set to true, notes
+		//                   that the field requires an argument e.g. field[arg].
+		//     "arg": (optional) if present, notes that the field can accept
+		//             an argument e.g. field[arg]. More precisely, the following
+		//             flags could be specified:
+		//             "isRequired": if true, the argument is required.
+		//             "isIndex": if true, the field is numeric.
+		//             "isKey": if true, the field is a string.
+		//             If "isRequired" is true, one between "isIndex" and
+		//             "isKey" must be true, to specify the argument type.
+		//             If "isRequired" is false, but one between "isIndex"
+		//             and "isKey" is true, the argument is allowed but
+		//             not required.
+		//     "display": (optional) If present, a string that will be used to
+		//                display the field instead of the name. Used in tools
+		//                like wireshark.
+		//     "desc": a string with a description of the field
+		// Example return value:
+		// [
+		//    {"type": "string", "name": "field1", "argRequired": true, "desc": "Describing field 1"}, [DEPRECATED 'argRequired' property]
+		//    {"type": "uint64", "name": "field2", "desc": "Describing field 2"},
+		//    {"type": "string", "name": "field3", "arg": {"isRequired": true, "isIndex": true,}, "desc": "Describing field 3"},
+		// ]
+		const char* (*get_fields)();
+		//
+		// Extract one or more a filter field values from an event.
+		// Required: no
+		// Arguments:
+		// - evt: an event struct produced by a call to next_batch().
+		//   This is allocated by the framework, and it is not guaranteed
+		//   that the event struct pointer is the same returned by the last
+		//   next_batch() call.
+		// - num_fields: the length of the fields array.
+		// - fields: an array of ss_plugin_extract_field structs. Each entry
+		//   contains a single field + optional argument as input, and the corresponding
+		//   extracted value as output. Memory pointers set as output must be allocated
+		//   by the plugin and must not be deallocated or modified until the next
+		//   extract_fields() call.
+		//
+		// Return value: A ss_plugin_rc with values SS_PLUGIN_SUCCESS or SS_PLUGIN_FAILURE.
+		//
+		ss_plugin_rc (*extract_fields)(ss_plugin_t *s, const ss_plugin_event *evt, uint32_t num_fields, ss_plugin_extract_field *fields);
+	};
+} plugin_api;
 
 //
-// Interface for a sinsp/scap extractor plugin
+// Small C interface that is passed down to libscap
```

Comment:
I'm not strong about this, but for my personal taste I would avoid adding an additional struct just for this and pass down to scap a `plugin_api*` instead. For the goal of reducing the plugin function visibility, I think an approach like the one of https://github.com/falcosecurity/libs/pull/213 will be the way to go once it will get merged. For the goal of storing state variables, I feel like this would not suffice anyways if we'll decide to support multiple source plugins. We can leave things as they are here for now, but we'll probably need to refactor this a little bit in the future!

Reply:
Yeah, i agree with you!  
I think that at the moment it makes no big difference and the needed refactor will be pretty small once #213 (or a similar design) is merged.
I am not sure how to currently store needed state variables too :)